### PR TITLE
UnitedDeployment: update subset

### PR DESCRIPTION
**Pull request by wu8685**
UnitedDeployment supports updating the subsets.

### CODE REVIEW

File: pkg/controller/uniteddeployment/uniteddeployment_update.go [hunk 1]
```diff
@@ -30,6 +30,49 @@ import (
...
+
+	expectedRevision := currentRevision
+	if updatedRevision != nil {
+		expectedRevision = updatedRevision
+	}
+
+	var needUpdate []string
+	for _, name := range exists.List() {
+		subset := (*nameToSubset)[name]
+		expectedPartition := (*nextPartitions)[name]
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/uniteddeployment_update.go]: Nit: we may not need expectedPartition variable as it is only used once.

File: pkg/controller/uniteddeployment/uniteddeployment_update.go [hunk 2]
```diff
@@ -30,6 +30,49 @@ import (
...
+	}
+
+	_, updateErr = util.SlowStartBatch(len(needUpdate), slowStartInitialBatchSize, func(index int) error {
+		cell := needUpdate[index]
+		subset := (*nameToSubset)[cell]
+		replicas := (*nextReplicas)[cell]
+		partition := (*nextPartitions)[cell]
+
+		klog.V(0).Infof("UnitedDeployment %s/%s needs to update Subset (%s) %s/%s with revision %s, replicas %d, partition %d", ud.Namespace, ud.Name, subsetType, subset.Namespace, subset.Name, expectedRevision.Name, replicas, partition)
+		updateIpsErr := r.subSetControls[subsetType].UpdateSubset(subset, ud, expectedRevision.Name, replicas, partition)
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/uniteddeployment_update.go]: Can we use a better name other than "updateIpsErr"?

File: pkg/controller/uniteddeployment/uniteddeployment_controller.go [hunk 3]
```diff
@@ -257,3 +259,115 @@ func (r *ReconcileUnitedDeployment) classifySubsetBySubsetName(ud *appsv1alpha1.
...
+
+func (r *ReconcileUnitedDeployment) calculateStatus(ud *appsv1alpha1.UnitedDeployment, nameToSubset *map[string]*Subset, nextReplicas, nextPartition *map[string]int32, currentRevision, updatedRevision *appsv1.ControllerRevision, collisionCount int32) *appsv1alpha1.UnitedDeploymentStatus {
+	newStatus := ud.Status
+
+	expectedRevision := currentRevision.Name
+	if updatedRevision != nil {
+		expectedRevision = updatedRevision.Name
+	}
+
+	var replicas int32
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/uniteddeployment_controller.go]: Do we need these four temporal variables?

File: pkg/controller/uniteddeployment/uniteddeployment_controller.go [hunk 4]
```diff
@@ -257,3 +259,115 @@ func (r *ReconcileUnitedDeployment) classifySubsetBySubsetName(ud *appsv1alpha1.
...
+	}
+
+	if newStatus.UpdateStatus == nil {
+		newStatus.UpdateStatus = &appsv1alpha1.UpdateStatus{}
+	}
+
+	newStatus.UpdateStatus.UpdatedRevision = expectedRevision
+	newStatus.UpdateStatus.CurrentPartitions = *nextPartition
+
+	if newStatus.UpdateStatus != nil && newStatus.UpdateStatus.UpdatedRevision != newStatus.CurrentRevision && newStatus.UpdatedReadyReplicas >= newStatus.Replicas {
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/uniteddeployment_controller.go]: This check newStatus.UpdateStatus != nil  seems to be unnecessary.

File: pkg/apis/apps/v1alpha1/uniteddeployment_types.go [hunk 5]
```diff
@@ -169,6 +169,9 @@ type UnitedDeploymentCondition struct {
 	// Status of the condition, one of True, False, Unknown.
 	Status corev1.ConditionStatus `json:"status,omitempty"`
 
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
```
Comment (Fei-Guo) [inline pkg/apis/apps/v1alpha1/uniteddeployment_types.go]: Is this field updated anywhere?

File: pkg/controller/uniteddeployment/uniteddeployment_update.go (hunk 1 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/uniteddeployment_update.go]: done

File: pkg/controller/uniteddeployment/uniteddeployment_update.go (hunk 2 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/uniteddeployment_update.go]: done

File: pkg/controller/uniteddeployment/uniteddeployment_controller.go (hunk 3 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/uniteddeployment_controller.go]: done

File: pkg/controller/uniteddeployment/uniteddeployment_controller.go (hunk 4 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/uniteddeployment_controller.go]: done

File: pkg/apis/apps/v1alpha1/uniteddeployment_types.go (hunk 5 above)
Comment (wu8685) [reply pkg/apis/apps/v1alpha1/uniteddeployment_types.go]: Removed this field and added the condition supports.

**Pull request merged**